src/stats/src/MetropolisAdjustedLangevinTK.C

```diff
@@ -0,0 +1,255 @@
+    double factor = 1./m_scales[i]/m_scales[i];
+    if ((m_env.subDisplayFile()        ) &&
+        (m_env.displayVerbosity() >= 10)) {
+      *m_env.subDisplayFile() << "In MetropolisAdjustedLangevinTK<V, M>::updateLawCovMatrix()"
+                              << ", m_scales.size() = " << m_scales.size()
+                              << ", i = "               << i
+                              << ", m_scales[i] = "     << m_scales[i]
+                              << ", factor = "          << factor
+                              << ": about to call m_rvs[i]->updateLawCovMatrix()"
+                              << ", covMatrix = \n" << factor*covMatrix // FIX ME: might demand parallelism
+                              << std::endl;
+    }
+    dynamic_cast<GaussianVectorRV<V, M> * >(m_rvs[i])->updateLawCovMatrix(factor*m_time_step*covMatrix);
+  }
+}
+
+template <class V, class M>
+bool
+MetropolisAdjustedLangevinTK<V, M>::setPreComputingPosition(const V & position, unsigned int stageId)
+{
+  if ((m_env.subDisplayFile()) && (m_env.displayVerbosity() >= 5)) {
+    *m_env.subDisplayFile() << "Entering MetropolisAdjustedLangevinTK<V, M>::setPreComputingPosition()"
+                           << ": position = " << position
+                           << ", stageId = "  << stageId
+                           << std::endl;
+  }
+
+  BaseTKGroup<V, M>::setPreComputingPosition(position, stageId);
+
+  if ((m_env.subDisplayFile()) && (m_env.displayVerbosity() >= 5)) {
+    *m_env.subDisplayFile() << "In MetropolisAdjustedLangevinTK<V, M>::setPreComputingPosition()"
+                           << ", position = "        << position
+                           << ", stageId = "         << stageId
+                           << ": preComputingPos = " << *m_preComputingPositions[stageId];
+    if (stageId < m_scales.size()) {
+      *m_env.subDisplayFile() << ", factor = " << 1./m_scales[stageId]/m_scales[stageId];
+    }
+    if (stageId < m_rvs.size()) {
+      const GaussianJointPdf<V, M>* pdfPtr = dynamic_cast< const GaussianJointPdf<V, M>* >(&(m_rvs[stageId]->pdf()));
+      *m_env.subDisplayFile() << ", rvCov = " << pdfPtr->lawCovMatrix(); // FIX ME: might demand parallelism
+    }
+    *m_env.subDisplayFile() << std::endl;
+  }
+
+  if ((m_env.subDisplayFile()) && (m_env.displayVerbosity() >= 5)) {
+    *m_env.subDisplayFile() << "Leaving MetropolisAdjustedLangevinTK<V, M>::setPreComputingPosition()"
+                           << ": position = " << position
+                           << ", stageId = "  << stageId
+                           << std::endl;
+  }
+
+  return true;
+}
+
+template <class V, class M>
+void
+MetropolisAdjustedLangevinTK<V, M>::clearPreComputingPositions()
+{
+  BaseTKGroup<V, M>::clearPreComputingPositions();
+}
+
+template <class V, class M>
+void
+MetropolisAdjustedLangevinTK<V, M>::setRVsWithZeroMean()
+{
+  queso_require_not_equal_to_msg(m_rvs.size(), 0, "m_rvs.size() = 0");
```

Comment:
The msg string here and elsewhere seems redundant with the built-in string.  Are users going to be perplexed to see "Assertion `m_rvs.size() != 0' failed." unless we can rescue them from confusion by continuing to explain, "m_rvs.size() = 0"?